### ISSUE TITLE
Jump on Link click and Visual Highlight

### ISSUE THREAD

**Issue by josiahseaman** [graph-genome/Schematize#6]
Links have two sides.  Jump to the side that's farthest from the current view window.
Depends on #4.  LinkColumn also jumps.  Do distance calculations from the center bin of the viewport.

**Comment by josiahseaman** [comment]
Still needs visual highlighting of the target link in order to close this issue.  I recommend having the target link extend all the way to the top of the screen and maybe do a color slide in the first second of render.  Any link you click on is likely a short truncated arrow.  Extending that arrow to full screen height should make it stand out from the crowd.

**Comment by fornof** [comment]
placeholder for design

**Comment by josiahseaman** [comment]
Robert Fornof will design what this user experience should look like.
- [ ] Mouseover should show pangenome coordinates.  Seperate issue already created?

**Example Use Case:**
Start at: 194 - 233,  There is a light green orphan arrow pointing down.  Clicking on this arrow jumps to pangenome coordinate start: 239 where there is a light green arrow pointing up.  These are the two ends which can be seen together if you set start and end to 207 - 258.

**Comment by fornof** [comment]
We propose a timeout of 5-10 seconds to highlight the previous entry point -time possibly configurable with a entry box.
![1Annotation 2020-04-06 202253](https://user-images.githubusercontent.com/1073801/78623766-97053980-7845-11ea-96d4-aeb95ce900ce.png)
note the highlight color may be different . Ideally this would fade out, but for a rough draft it can just stay highlighted and cut out. 
![2Annotation 2020-04-06 202456](https://user-images.githubusercontent.com/1073801/78623773-979dd000-7845-11ea-9e31-375e4e8cae04.png)
 this is the end point, ideally it would be highlighted

OR we could have it where it highlights and stays highlighted until the next highlight - as the lastHighlighted feature. color it light blue or something.

**Comment by josiahseaman** [comment]
TODO: Post design spec to this issue.

**Comment by josiahseaman** [comment]
Robert Fornof worked on this issue but hasn't made a PR, so the diff will need to be manually added into a commit and PR.  

[Download Link Here.](https://virtualbiohac-xt62674.slack.com/files/U010MREFH2P/F011PG7B8PP/diffhighlight.diff?origin_team=T010K0KRLTF&origin_channel=G01177M7AMQ)

You can see a [Video of the Interaction Here](https://virtualbiohac-xt62674.slack.com/files/U010MREFH2P/F011MU6H6SW/2020-04-11_15-15-50.mkv?origin_team=T010K0KRLTF&origin_channel=G01177M7AMQ).

```diff
diff --git a/src/App.js b/src/App.js
index 1dc62e8..4f63a20 100644
--- a/src/App.js
+++ b/src/App.js
@@ -91,7 +91,7 @@ class App extends Component {
             });
         }
     }
-    recalcXLayout(){
+     recalcXLayout(){
         console.log("recalcXLayout")
         const sum = (accumulator, currentValue) => accumulator + currentValue;
         let columnsInComponents = this.schematic.components.map(component =>
@@ -111,8 +111,9 @@ class App extends Component {
         this.props.store.updateTopOffset(parseInt(top));
     }
 
-    recalcY(){
+     recalcY(){
         //forceUpdate() doesn't work with callback function
+        console.log('recalc y')
         this.setState({highlightedLink: null}); //nothing code to force update.
     }
 
@@ -135,7 +136,7 @@ class App extends Component {
         return maxNumberRowsInOneComponent;
     }
 
-    visibleHeight(){
+     visibleHeight(){
         if (this.props.store.useVerticalCompression || !this.compressed_row_mapping) {
             // this.state.schematize.forEach(value => Math.max(value.occupants.filter(Boolean).length, maxNumberRowsInOneComponent));
             return (this.maxNumRowsAcrossComponents + 2.5) * this.props.store.pixelsPerRow;
@@ -201,11 +202,11 @@ class App extends Component {
             </>
         )
     }
-
+    
 
     renderLinkColumn(schematizeComponent, i, firstDepartureColumn, j, linkColumn) {
         let xCoordArrival = this.leftXStart(schematizeComponent,i, firstDepartureColumn, j);
-        let localColor = stringToColor(linkColumn, this.state.highlightedLink);
+        let localColor = '#0000ff' // stringToColor(linkColumn, this.state.highlightedLink);
         return <LinkColumn
             store={this.props.store}
             key={"departure" + i + j}
@@ -262,13 +263,14 @@ class App extends Component {
 
     render() {
         console.log("Start render");
+        console.log(this.props.store)
         return (
             <>
                 <ControlHeader store={this.props.store}/>
                 <Stage
                     x={this.props.store.leftOffset}
                     y={this.state.buttonsHeight} //removed leftOffset to simplify code.  Relative coordinates are always better.
-                    width={this.state.actualWidth + 60}
+                    width={this.state.actualWidth }
                     height={this.props.store.topOffset + this.visibleHeight()}>
                     <Layer ref={this.layerRef}>
                         {this.renderSchematic()}
@@ -285,10 +287,11 @@ class App extends Component {
                               x={-20}
                               y={-this.state.buttonsHeight+15}
                               width={this.state.actualWidth+20}
-                              height={this.props.store.topOffset+this.state.buttonsHeight}
-                              fill="white"
+                              height= {this.props.store.topOffset+this.state.buttonsHeight}
+                              fill="gray"
                               />
-                            {this.renderSortedLinks()}
+                              
+                             {this.renderSortedLinks()} 
                         </Layer>
                   </Stage>
                 <NucleotideTooltip store={this.props.store}/>
diff --git a/src/LineAdjustment.js b/src/LineAdjustment.js
new file mode 100644
index 0000000..ff04553
--- /dev/null
+++ b/src/LineAdjustment.js
@@ -0,0 +1,37 @@
+import Konva from 'konva';
+
+class LineAdjustment{
+
+    constructor({config={}}){
+        this.defaultTimeout = config.defaultTimeout||8
+        this.strokeColor = config.strokeColor|| '#00ff00'
+        this.strokeWidth = config.strokeWidth|| 10
+    }
+    // this handles a highlight given a click event. it adds a timeout and resets things 
+    //back to how they were after timeout
+    handleHighlight(event){
+        const attrsCopy = Object.assign({},event.target.attrs)
+        const timeout = setTimeout(this.backToNormal.bind(this,event,attrsCopy),this.defaultTimeout*1000)
+        event.target.to({
+            duration: 1,
+            easing: Konva.Easings.EaseIn,
+            // scaleX: 1.5,
+            // scaleY: 1,
+            stroke: '#22ff00',
+            strokeWidth: 10,
+            shadowOffsetX: 5,
+            shadowOffsetY: 5
+          });
+    
+    }
+
+    // given an event to modify a target of from the setTimeout above, it puts things back to normal
+    backToNormal(event, eventCopy){
+        let data = eventCopy
+        data.duration = 3
+        data.easing = Konva.Easings.ElasticEaseOut
+        console.log('target', data)
+        event.target.to(data);
+    }
+}
+export default LineAdjustment
\ No newline at end of file
diff --git a/src/LinkArrow.js b/src/LinkArrow.js
index 86f4b81..11d3b53 100644
--- a/src/LinkArrow.js
+++ b/src/LinkArrow.js
@@ -1,6 +1,6 @@
 import React  from 'react';
 import {Arrow} from 'react-konva';
-
+import LineAdjustment from '../src/LineAdjustment'
 function isInfinity(n){
     return !Number.isFinite(n);
 }
@@ -11,6 +11,7 @@ class LinkArrow extends React.Component {
         super(props);
         this.arrowXCoord = null;
         this.points = [];
+        this.lineAdjustment = new LineAdjustment({})
         this.handleMouseOut = this.handleMouseOut.bind(this);
         this.handleMouseOver = this.handleMouseOver.bind(this);
     }
@@ -97,9 +98,9 @@ class LinkArrow extends React.Component {
         this.props.updateHighlightedNode(null)
     };
     handleClick = (event) => {
+        this.lineAdjustment.handleHighlight(event)
         /*Jump on Link click rough draft. Detects which end is closest to the view and
         jumps to the other side. TODO: Still needs visual highlighting at destination.*/
-        console.log("Click", event, this.props.link);
         //find middle position of viewport
         let width = this.props.store.getEndBin() - this.props.store.getBeginBin();
         let mid_bin = this.props.store.getBeginBin() + width /2;
```

### Previous Conversation:
**Josiah:** So it looks like you ended up modifying the codebase instead of build your own demo from scratch.  Good job diving in.  I've got a lot of Pull Requests to integrate, but I'd like to talk about the design more than the code.  I like the color fades, you were also planning on using stroke?
is stroke meant to be permanent? I was also surprised by the x-axis slide animation but I kind of like it.  Is the idea that the corresponding arrow is sliding in from off screen?  To indicate if it's pair is to the left or to the right?  
**Robert:** I made a class to change this and changed outside, then moved the code base inside. I thought a pull request was for a different branch into another branch?
Stroke is what I used to make it bigger. The ElasticOut can be probably just EaseOut - so there's no wobbling.
The arrows surprised me too. I have a feeling some arrows are reused.
stroke is temporary. It highlights to prove that is where your destination is, and x-slides so you may know where you came from - I would need someone with domain knowledge to check these over but it seems to make sense.  
I'm not sure that is the cleanest diff, let me know if it needs to be edited.

**Comment by AndreaGuarracino** [comment]
I have some suggestions (perhaps not all new):
a) when the mouse is on a Link or a LinkColumn, all the others (Link and LinkColumns) turn gray and fade out;
b) when the mouse leaves the Link/LinkColumn, all the colors return.

![image](https://user-images.githubusercontent.com/62253982/80822178-6ad9a000-8bda-11ea-8a8a-f8c816d282b9.png)

These two behaviours are **delayed**: a) happens only if the mouse remains on the same Link for at least X ms (parameter to tune); b) happens only if the mouse leaves the last Link for at least Y ms (parameter to tune). In this way, we avoid lots of rendering while we move the mouse on the Links and LinkColumns (we don't have to do what we are not sure the user wants).

Moreover, when you **click** on a Link, the visualized window changes, the clicked Links remains colored and all the other turn grey and fade out: this situation remains for Z ms (parameter to tune) also if the mouse leaves the clicked Link, or until you click another Link.

Finally, when a Link is selected or clicked, the corrisponding LinkColumns are visualized with a stroke to explicit them more, but I can remove this behaviour if you think it is unnecessary.

About the window-change when you click on a Link, I think there is still work to do, and I would avoid to change the visualized window if the clicked Link is already enterily visualized in the current window. But this is another issue.

**Comment by subwaystation** [comment]
For me, all propositions make sense ;)
For the parameters to tune, maybe take a look at http://courses.ischool.berkeley.edu/i213/s99/Readings/p181-card.pdf and http://yusufarslan.net/sites/yusufarslan.net/files/upload/content/Miller1968.pdf. One thing they state is that a response time of about 1.0 seconds presents the limitation of the user’s flow of thought.

**Comment by AndreaGuarracino** [comment]
At the moment, my brain feels quiete good with a delay of 600 ms, but when the entire performance will improve, we should re-tune them.

Known issues:
- when a new FASTA file is read for the first time, very often the file content "arrives" after the nucleotide visualization (which will obviously be empty), so you need to re-render the scenario to see the sequence which is ready in memory. This problem should be tackled togheter with the event management to improve the overall perfomance.
- when you click on a Link, the window-shift doesn't seem to work precisly every time, and I would avoid to change the visualized window if the clicked Link is already enterily visualized in the current window.
- when you chose a Column Widht < 10, some nucleotides disappear. Maybe, this issue is related to this one #80.

**Comment by josiahseaman** [comment]
A screenshot of the Column Width < 10 issue.  Funny but easily solved by making a single check globally as opposed to per letter.

![image](https://user-images.githubusercontent.com/2256148/81085811-ef673f80-8eb4-11ea-9825-2f1424613d8b.png)
